astroid/cache.py

```diff
@@ -0,0 +1,100 @@
+""" Caching utilities used in various places."""
+
+from collections import OrderedDict
+from weakref import WeakSet
+
+import wrapt
+
+LRU_CACHE_CAPACITY = 128
+
+
+class LRUCache:
+    """An LRU cache that keeps track of its instances."""
+
+    instances = WeakSet()
+
+    def __init__(self):
+        self.cache = OrderedDict()
+        LRUCache.instances.add(self)
+
+    def __setitem__(self, key, value):
+        self.cache[key] = value
+
+        if len(self.cache) > LRU_CACHE_CAPACITY:
+            self.cache.popitem(last=False)
+
+    def __getitem__(self, key):
+        if key in self.cache:
+            self.cache.move_to_end(key)
+
+        return self.cache[key]
+
+    def __contains__(self, key):
+        return key in self.cache
+
+    def clear(self):
+        self.cache.clear()
+
+    @classmethod
+    def clear_all(cls):
+        """Clears all LRUCache instances."""
+        for cache in cls.instances:
+            cache.clear()
+
+
+def lru_cache(arg=None):
```

Comment:
I wonder if we should rename this to `lru_cache_astroid` just so we don't create any confusion with the `functools` one.

Reply:
Renamed it to `@lru_cache_astroid` following your suggestion.